protocol/x/vault/keeper/withdraw.go

```diff
@@ -42,3 +44,100 @@ func (k Keeper) ValidateWithdrawFromVault(
 
 	return nil
 }
+
+func (k Keeper) RedeemShares(
+	ctx sdk.Context,
+	vaultId types.VaultId,
+	owner string,
+	sharesToRedeem types.NumShares,
+) (quantumsToWithdraw *big.Int, err error) {
+	// **** Validate ****
+	// Validate that `sharesToRedeem` is positive.
+	sharesToRedeemInt := sharesToRedeem.NumShares.BigInt()
+	if sharesToRedeemInt.Sign() <= 0 {
+		return nil, errors.Wrapf(
+			types.ErrInvalidWithdrawalAmount,
+			"Shares to withdraw must be positive",
+		)
+	}
+	totalShares, vaultExists := k.GetTotalShares(ctx, vaultId)
+	totalShareInt := totalShares.NumShares.BigInt()
+	ownerShares, ownerSharesExist := k.GetOwnerShares(ctx, vaultId, owner)
+	ownerSharesInt := ownerShares.NumShares.BigInt()
+	// Validate that vault and owner shares exist.
+	if !vaultExists || !ownerSharesExist {
+		return nil, errors.Wrapf(
+			types.ErrOwnerShareNotFound,
+			"VaultId: %v, Owner: %v",
+			vaultId,
+			owner,
+		)
+	}
+	// Validate that owner shares and total shares are positive.
+	if ownerSharesInt.Sign() <= 0 || totalShareInt.Sign() <= 0 {
+		return nil, errors.Wrapf(
+			types.ErrNegativeShares,
+			"Owner/Total shares must be positive. OwnerShares: %v, TotalShares: %v",
+			ownerSharesInt,
+			totalShareInt,
+		)
+	}
+	// Validate that
+	// 1. `sharesToRedeem` are less than or equal to the owner shares and
+	// 2. owner shares are less than or equal to the total vault shares.
+	if ownerSharesInt.Cmp(sharesToRedeemInt) < 0 ||
```

Comment:
i thought in this case, we simply withdraw all owner shares rather than failing the tx?